### ISSUE TITLE
Spotify endpoints not working for me

### ISSUE THREAD

**Issue by blake06** [jishi/node-sonos-http-api#137]
Hello,

Thanks for the great work!  I'm using this for my Sonos speakers, but I am unable to get the Spotify endpoints to work.

Line 21 in spotify.js never passes the check (I added in some console.log statements to test it), as it appears that player.coordinator.addURIToQueue is always returning false to the callback.

Is there something that I might be missing that is needed to get those Spotify endpoints functioning?

Thanks!
Blake

**Comment by jishi** [comment]
Where do you live, and what TYPE and UN for spotify do you see under the following page:

http://ip_to_player:1400/status/accounts

Could you also log the following variables before line 21, `uri` and `metadata` and tell me what those contains? They should not contain any sensitive information regarding your account, if they do, just replace it with xxxx or something.

**Comment by blake06** [comment]
Thanks for the quick response!

I'm in the United States.   I see Type: 3079 and UN: X_#Svc3079-0-Token

I'll add in the logging for uri and metadata and get back to you in just a bit.

**Comment by blake06** [comment]
uri: x-sonos-spotify:spotify%3Atrack%3A0Zj65VBQqXiRMvO2pjhOMy?sid=9&flags=32&sn=1

metadata: <DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"><item id="00030020spotify%3Atrack%3A0Zj65VBQqXiRMvO2pjhOMy" restricted="true"><upnp:class>object.item.audioItem.musicTrack</upnp:class><desc id="cdudn" nameSpace="urn:schemas-rinconnetworks-com:metadata-1-0/">SA_RINCON2311_X_#Svc2311-0-Token</desc></item></DIDL-Lite>

**Comment by blake06** [comment]
I replaced the 2311 with 3079 in the metadataTemplate variable on Line 1, and it's working now!

**Comment by jishi** [comment]
Nice to know. Yeah, I figured 2311 was related to EU only.

**Comment by blake06** [comment]
One more question.  It looks like the spotify endpoints will only play a track.  Do you know if it would be possible to play a Spotify URI for an album, an artist, or a playlist?

For example:

spotify:album:4fNobiJPOvXUFVypz4yzZV
spotify:user:stephenjamesgomez:playlist:3gwlb4ediWAz9odZD0IJkW
spotify:artist:2qc41rNTtdLK0tV3mJn2Pm

**Comment by jishi** [comment]
Not at the moment. I think it is possible to queue whole albums and playlists using their corresponding URIs, but I haven't investigated it further. As I said, this is highly experimental. Right now I'm focusing on a rewrite of the underlying framework since it is long overdue.

**Comment by blake06** [comment]
No problem.  Thanks again!

**Comment by blake06** [comment]
Just FYI that I was able to get it working for playlists and for albums, but couldn't get the artist URI working.   Attached is what I used in case it's at all helpful for future versions.
[spotifyplaylist.js.zip](https://github.com/jishi/node-sonos-http-api/files/105224/spotifyplaylist.js.zip)

**Comment by jishi** [comment]
Well, what do you expect it to play of you send in an artist?

**Comment by blake06** [comment]
Honestly, no idea.  I was thinking maybe it would be a playlist-like list of all the songs by that artist.  But maybe that's not at all what the Artist URI is used for. 

But getting the playlists and albums working is perfect for what I need.

I'm currently using this:  https://github.com/rgraciano/echo-sonos

I've updated the Alexa Skill to allow me to tell my Echo to search for a playlist, track, or an album, and it uses the Spotify API to search, and passes the 1st search item to sonos using your spotify action (for track searches) or the spotifyplaylist action (for playlist and album searches).

It's working great.   I'll be sending you a donation via paypal as a small token of thanks.

**Comment by jishi** [comment]
I think Spotify just uses that uri to locate an artist in their backend. There is probably another uri for top tracks for an artist, maybe it is possible to find it using their web api?

Appreciate the donation!

**Comment by jishi** [comment]
You seem to need to access them track by track, https://api.spotify.com/v1/artists/43ZHCT0cAZBISjO8DG9PnE/top-tracks?country=SE, but doable.

**Comment by blake06** [comment]
Yep, was just reading through that API call documentation.  Looks like you can get up to 10 top tracks, and I could loop through and queue them each.   I might play with that in the future.

**Comment by scottlow** [comment]
@blake06 Do you mind sharing your Alexa skill for echo-sonos? And is the code you used for node-sonos included in the zip you posted above? I was just going to sit down and write what you described myself, but figured I'd ask if you already have it done :)

Thanks!

**Comment by blake06** [comment]
@scottlow 

I'm using https://github.com/rgraciano/echo-sonos for the Alexa Skill.

But I modified the Intent Schema and the Sample Utterances, and I set up some Custom Slot Types.

I've attached a zip file with the Intent Schema I'm using and the Sample Utterances.

For the Custom Slot Types, I have one for PRESETS with the name of all my presets that I created in node-sonos.  I have one for ROOMS with the names of all my Sonos rooms.  And then I have one for TOGGLES with on and off as the values.

I've also updated the Lambda code for echo-sonos.  I've attached the src.zip file for it.

For node-sonos, all I modified was the lib/actions/spotify.js (to update the sections that needed changed for it to work in the US), and I created lib/actions/spotifyplaylist.js which is included in the zip file from up earlier in this thread.

[src.zip](https://github.com/jishi/node-sonos-http-api/files/125036/src.zip)



[EchoSonosUtterancesIntents.zip](https://github.com/jishi/node-sonos-http-api/files/125035/EchoSonosUtterancesIntents.zip)
